src/psyclone/parse/file_info.py

```diff
@@ -77,27 +143,286 @@ def filename(self):
+            ) -> Fortran2003.Program:
+        """Returns the fparser Fortran2008 representation of the source code.
+
+        :param save_to_cache: Cache is updated if fparser was
+            not loaded from cache.
+        :param verbose: Produce some verbose output
+
+        :returns: fparser representation.
+
+        :raises FileInfoFParserError: if fparser had issues
+        """
+        if self._fparser_tree is not None:
+            return self._fparser_tree
+
+        if verbose:
+            print(f"- Source file '{self._filename}': " f"Running fparser")
+
+        source_code = self.get_source_code()
+        assert self._source_code_hash_sum is not None
+
+        # Check for cache
+        self._cache_load(verbose=verbose)
+
+        if self._cache_data_load is not None:
+            if self._cache_data_load._fparser_tree is not None:
+                if verbose:
+                    print(
+                        f"  - Using cache of fparser tree with hashsum"
+                        f" {self._cache_data_load._source_code_hash_sum}"
+                    )
+
+                # Use cached version
+                self._fparser_tree = self._cache_data_load._fparser_tree
+                return self._fparser_tree
+
+        try:
+            reader = FortranStringReader(
+                source_code, include_dirs=Config.get().include_paths
+            )
+            parser = ParserFactory().create(std="f2008")
+            self._fparser_tree = parser(reader)
+
+        except Exception as err:
+            raise FileInfoFParserError(
+                "Failed to get fparser tree: " + str(err)
+            ) from err
+
+        # We directly call the cache saving routine here in case that the
+        # fparser tree will be modified later on.
+        if save_to_cache:
+            self._cache_save(verbose=verbose)
+
+        return self._fparser_tree
+
+    def get_psyir_node(self, verbose: bool = False) -> FileContainer:
+        """Returns the psyclone FileContainer of the file.
+
+        :param verbose: Produce some verbose output
+
+        :returns: psyclone file container node.
+
+        """
+        if self._psyir_node is not None:
+            return self._psyir_node
+
+        # Check for cache
+        self._cache_load(verbose=verbose)
+
+        if self._cache_data_load is not None:
+            if self._cache_data_load._psyir_node is not None:
+                # Use cached version
+                if verbose:
+                    print("  - Using cache of fparser tree")
+
+                self._psyir_node = self._cache_data_load._psyir_node
+                return self._psyir_node
+
+        if verbose:
+            print(f"  - Running psyir for '{self._filename}'")
+
+        # First, we get the fparser tree
+        # TODO #2786: use 'save_to_cache=False' if TODO is resolved
+        fparse_tree = self.get_fparser_tree(
+                verbose=verbose,
+                save_to_cache=True
```

Comment:
This needs to default to False for the moment. You could expose it (for development purposes) as a module variable or a class variable.

Reply:
You're right. Thanks for spotting this bug.
(But it wouldn't be cached if cache is disabled. I renamed this argument to `save_to_cache_if_cache_active` to reflect this better)

Reply:
This has to be `True`. The "idea" is as follows:

If loading only the `fparser tree` by calling `get_fparser_tree` from some other program using psyclone, we want to potentially cache the fparser tree if it's not yet cached.

If loading the `PSyIR`, we don't want to cache the fparser tree directly after loading it, but wait until we also loaded the `PSyIR` to store both together to the cache file. Then, if caching would be active, `save_to_cache_if_cache_active=False` to delay updating the cache.
However, we don't have the caching of `PSyIR`, yet.
Hence, we update the cache even if we just loaded the fparser by setting this to `True`.
I add a comment on the TODO to this.

As pointed out above, if the cache is not active, setting this to `True` won't write the cache file.